test/test-lint.cpp

```diff
@@ -2811,6 +2811,201 @@ TEST(
                           ERROR_TYPE_FIELD(error_use_of_undeclared_variable,
                                            name, span_matcher(use_after))));
 }
+
+TEST(test_lint_eval, disable_variable_lookup_in_presence_of_eval) {
+  const char8 use_eval[] = u8"eval";
+  const char8 use[] = u8"x";
+  {
+    // eval("var x = 42");
+    // x;
+    // x = 10;
+    error_collector v;
+    linter l(&v);
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // {
+    //   eval("var x = 42");
+    // }
+    // x;
+    // x = 10;
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_block_scope();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_exit_block_scope();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // eval("var x = 42");
+    // function f() {
+    //   x;
+    //   x = 10;
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // function f() {
+    //   x;
+    //   x = 10;
+    // }
+    // eval("var x = 42");
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // function f() {
+    //   x;
+    //   x = 10;
+    //   eval("var x = 42;");
+    //   x;
+    //   x = 10;
+    // }
```

Comment:
I think this test deserves some explanation. For example:

> In this case, we do know that the first references to `x` will error. However, the current algorithm is conservative, because variable references before `eval` might be executed after the `eval` (e.g. if there is a loop and the earlier references are conditional).

Reply:
Should I put a comment for this test?

Reply:
Sure